Track original source location in lib_info/dune_package

andreypopp opened the pull request:
This allows to communicate original source locations between installed (dune) libraries so we can construct `.merlin` files with `S` directives pointing right to them for installed libs. Thus allowing merlin to see original sources.

This is useful when lib is opam-pinned or esy-linked. I think we can set `orig_src_dir` to `None` for all other installed packages (based on profile?).

andreypopp committed:
Track original source location in lib_info/dune_package

This allows to communicate original source locations so we can construct
`.merlin` files with `S` directives pointing right to them for installed
libs.

This is useful when lib is opam-pinned or esy-linked. I think we can set
`orig_src_dir` to `None` for all other installed packages (based on
profile?).

Signed-off-by: Andrey Popp <8mayday@gmail.com>

andreypopp commented on src/lib_info.ml:
```diff
@@ -84,6 +85,12 @@ let user_written_deps t =
+    match Path.drop_build_context dir with
+    | None -> None
+    | Some src_dir ->
+      Some Path.(of_string (to_absolute_filename (in_source (to_string src_dir))))
```
This is weird and maybe can be done easier but `Path` is confusing for me at the moment.

diml replied:
`in_source (to_string src_dir)` should be the identity, you can replace it by `src_dir`

rgrinberg commented:
The idea is fine, but how do we control which sources to use in the .merlin file? If a package is installed via opam, then the original source dir will be ephemeral.

andreypopp commented:
I thought maybe if profile is set to “release” (-p is passed) we don’t
populate “orig_src_dir” in dune package metadata.

On Thu, 10 Jan 2019 at 19:45, Rudi Grinberg <notifications@github.com>
wrote:

> The idea is fine, but how do we control which sources to use in the
> .merlin file? If a package is installed via opam, then the original source
> dir will be ephemeral.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ocaml/dune/pull/1750#issuecomment-453165895>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAB3glYuDiW0e-A-boEyUD1GUD3DyZJtks5vB24dgaJpZM4Z5bvw>
> .
>
-- 
Andrey Popp / 8mayday@gmail.com

diml commented:
What about controlling this behaviour with a command line switch/environment variable? In practice, we probably don't want to store these paths by default as it will hurt reproducibility, but esy could enable it via the environment variable.

andreypopp commented:
> What about controlling this behaviour with a command line switch/environment variable?
> In practice, we probably don't want to store these paths by default as it will hurt reproducibility, but esy could enable it via the environment variable.

Won't such option be overly specific? Doesn't dune already have `dev` vs `release` build modes — we can generate such paths only in `dev` mode. I think that will be useful for opam-pinned packages as well (they can define build command optionally using `dev` profile if `pinned` is `true` in opam).

andreypopp committed:
Only emit orig_src_dir to dune-package in dev profile

Signed-off-by: Andrey Popp <8mayday@gmail.com>

andreypopp committed:
Sanitize output in blackbox-tests to replace testcase root path

This replaces test case root with $TESTCASE_ROOT token against which we
can match in expectations.

Signed-off-by: Andrey Popp <8mayday@gmail.com>

andreypopp committed:
Update tests due to changes to orig_src_dir

Signed-off-by: Andrey Popp <8mayday@gmail.com>

andreypopp committed:
Check that we don't emit orig_src_dir in release profile

Signed-off-by: Andrey Popp <8mayday@gmail.com>

andreypopp commented:
I proceeded with not emitting `orig_src_dir` in non `dev` profile (related: should we make profile `type kind = Dev | Release and profile = Profile of kind | Custom of kind * name`?).

I had to add a sanitisation step for blackbox test which replaces test's root path with `$TESTCASE_ROOT` token against which we could match in test expectations.

andreypopp commented on test/blackbox-tests/test-cases/dune-package/run.t:
```diff
@@ -1,7 +1,44 @@
+    (modules ((name Y) (obj_name c__Y) (visibility public) (impl)))
+    (wrapped true)))
+
+Build with "release" profile
```
This is a new test case but diff doesn't make it obvious.

andreypopp commented on test/blackbox-tests/test-cases/variants/run.t:
```diff
@@ -330,6 +332,8 @@ Include variants and implementation information in dune-package
    (kind normal)
    (virtual)
+   (orig_src_dir
+    $TESTCASE_ROOT/dune-package-info/vlib)
```
I'm not sure it's ok this breaks on the next line here.

andreypopp committed:
Update CHANGES

Signed-off-by: Andrey Popp <8mayday@gmail.com>

rgrinberg commented:
It just occurred to me that it's possible to implement this feature without touching dune-package at all.

The only reason why dune doesn't know that some external packages are actually local is because they get filtered out by `-p`. Well it shouldn't be hard to change dune to remember which library definitions it has skipped.

andreypopp commented:
> The only reason why dune doesn't know that some external packages are actually local is because they get filtered out by -p. Well it shouldn't be hard to change dune to remember which library definitions it has skipped.

What does it mean "actually local" — being a part of the same workspace? Then this isn't the case always in esy: it allows you to link packages from outside:
```
{
  "resolutions": {
    "@opam/dune": "link-dev:../path/to/dune/checkout"
  }
}
```

rgrinberg commented:
I see. So my suggestion wouldn't work after all. Would it be possible for esy to just pass the linked packages to dune somehow? Changing dune-package is a touchy issue for us, so I'd prefer that we refrained from doing so unless we really had no other way.

andreypopp commented:
> Changing dune-package is a touchy issue for us, so I'd prefer that we refrained from doing so unless we really had no other way.

Instead of changing `dune-package` we could place `dune-package.dev` with dev
related metadata (`orig_src_path` only at first but there can be potentially
more in the future).

`OCAMLPATH` (findlib) config is already used to communicate about installed
OCaml artifacts in the esy project (or system/switch wide in case of opam). I
think adding additional package dev metadata there makes sense. But I'm open for suggestions.

andreypopp commented:
> It seems to me that using an environment variable is the best option here. It's
> true that this is a very specific setting, but on the other hand most of the
> time storing the original source directory doesn't make sense. Even if we store
> it in a separate .dev file, it still doesn't feel right to install such files
> with temporary paths.

Ok, makes sense.

What would be the name/value of such environment variable? Something like (using
s-expressions):
```
DUNE_LIB_SOURCES='(lwt "/path/to/lwt" fmt "/path/to/fmt")'
```

andreypopp commented:
Hm... on a second thought I'm not sure how esy would be able to construct such
environment variable as it doesn't know about ocamlfind libraies which packages
are shipping.

andreypopp commented:
Ok, sorry for the noise, I think now you meant controlling that behaviour as passing a boolean flag via environment. I agree, that could work. What would be the name of such environment variable? `DUNE_STORE_ORIG_SOURCE="true"` or something else?

diml commented:
Yep, that's what I meant. I suggest `DUNE_STORE_ORIG_SOURCE_DIR` to be more precise.

andreypopp commented:
@diml thanks! Will update this PR with that.

andreypopp committed:
Store orig_src_dir based command line flag

Use `--store-orig-source-dir` command line flag (or
`DUNE_STORE_ORIG_SOURCE_DIR` env var) as suggested by @diml.

Signed-off-by: Andrey Popp <8mayday@gmail.com>

andreypopp committed:
Update blackbox test cases due to --store-orig-source-dir

Signed-off-by: Andrey Popp <8mayday@gmail.com>

andreypopp committed:
Update CHANGES

Signed-off-by: Andrey Popp <8mayday@gmail.com>

andreypopp commented:
This is ready for review.

andreypopp commented:
@rgrinberg the only reason I can think of is that users will need to configure `dune-workspace` if they want to merlin to work correctly in multi package environments. I'd prefer esy to be able to do this automatically somehow but generating `dune-workspace` isn't very convenient.

rgrinberg commented:
>  I'd prefer esy to be able to do this automatically somehow but generating dune-workspace isn't very convenient.

It will not be necessary [1](https://github.com/ocaml/dune/pull/1711#issuecomment-452095356)

andreypopp commented:
@rgrinberg I guess before that we could manually set `dune-workspace` file. Ok, is the stanza `(store_orig_source_dir)` ok for `dune-workspace` then?

diml commented:
Hold on, one big advantage of using specific environment variables is that it makes composition easy. For instance, I believe that it is now generally accepted that `OCAMLPARAM` is a pain to use and that specific variables would have been better.

@rgrinberg why is it annoying to introduce new environment variables?

jordwalke commented:
I have to chime in here to say that some kind of env var does seem better to me (than the alternative) because it's something package authors can't know to configure - only the package manager knows what's linked for the purposes of dev vs. release. I don't have many opinions on exactly what the env var looks like, but this is the one remaining feature blocking a very great monorepo workflow for esy and dune which allows you to "absorb" external packages into your local monorepo temporarily while developing.

andreypopp commented:
Any decision on how we should proceed with this?

diml commented on test/blackbox-tests/sanitize.ml:
```diff
@@ -0,0 +1,31 @@
+open Printf
```
We can do this in `cram.mll` directly. We already sanitise a few things there such as OS-specific extensions.

diml commented on test/blackbox-tests/test-cases/dune-package/run.t:
```diff
@@ -28,3 +28,75 @@
+
+Build with "--store-orig-source-dir" profile
+  $ dune build --store-orig-source-dir
+  $ cat _build/install/default/lib/a/dune-package
```
I'd put `| grep -A 1 orig_src_dir`, just to focus the test on the relevant part. Otherwise the reader has to manually scan the output to check that the test is successful.

diml commented:
Introducing new environment variables seems fine to me. Let's proceed with using a specific environment variable.

andreypopp committed:
Cleanup orig_src_dir from stanza

Signed-off-by: Andrey Popp <8mayday@gmail.com>

andreypopp committed:
test/blackbox-tests: replace sanitize.ml with cram postprocessing

Signed-off-by: Andrey Popp <8mayday@gmail.com>

andreypopp committed:
test/blackbox-tests/test-cases/dune-package: narrow down output

Signed-off-by: Andrey Popp <8mayday@gmail.com>

andreypopp commented:
Thanks for the feedback @diml, I believe I've addressed all points.

rgrinberg commented:
The reason why I wanted to started with a workspace option is that it avoids habit to remember which vars can be overridden via the environment.

On the other hand, there's nothing wrong with an individual env var even if there a workspace the option.

diml commented:
Yh that's true. In this case, I expect that this variable will be set by another tool most of the time.

diml commented on test/blackbox-tests/test-cases/merlin-tests/run.t:
```diff
@@ -13,9 +13,9 @@
   B ../_build/default/lib/.foo.objs/public_cmi
   S $LIB_PREFIX/lib/bytes
   S $LIB_PREFIX/lib/findlib
+  S $TESTCASE_ROOT/lib
```
Does this mean that we now use the absolute filename even for local libraries? I vaguely remember that we use to do that and changed it for some reason. Could you make sure we use `orig_src_dir` for external libraries only?

andreypopp replied:
Agree, sorry for bundling this meaningless change, fixed now.

diml commented:
@andreypopp thanks, just one last point I missed during the first review, otherwise it looks good.

andreypopp committed:
Do not store orig_src_dir for local dune libs

Signed-off-by: Andrey Popp <8mayday@gmail.com>

andreypopp commented:
Thanks for another round of feedback @diml, I believe I've address these points too. Also rebased on top of the current `master` to resolve conflicts.

diml commented:
Thanks, LGTM

andreypopp commented:
Thanks!
